Feat/sdl 0138 audio pass thru array

shoamano83 opened the pull request:
Fixes https://github.com/smartdevicelink/sdl_core/issues/2024

This PR is **ready** for review.

### Risk
This PR makes **no** API changes.

### Testing Plan
Unit tests are added / updated to cover the scenario.

### Summary
- This PR updates HMI_API.xml according to the proposal https://github.com/smartdevicelink/sdl_evolution/blob/master/proposals/0138-hmi-audiopassthru-capability.md.
- Core's logic is updated to choose "audioPassThruCapabilitiesList" param in UI.GetCapabilities if it is available.
- Core's implementation is updated to support "audioPassThruCapabilities" entry of an array in hmi_capabilities.json.
- "audioPassThruCapabilities" entry in hmi_capabilities.json is changed to array.

### Changelog
##### Enhancements
* Update HMI_API.xml to add audioPassThruCapabilitiesList 
* Update audioPassThruCapabilities in hmi_capabilities.json to array 
* Use utility method for reading pcmStreamCapabilities also 

##### Bug Fixes
* None

### CLA
- [X] I have signed [the CLA](https://docs.google.com/forms/d/e/1FAIpQLSdsgJY33VByaX482zHzi-xUm49JNnmuJOyAM6uegPQ2LXYVfA/viewform)

shoamano83 commented:
Hi @JackLivio , @jacobkeeler , please kindly review. Thanks.

shoamano83 commented:
Rebased onto latest `develop`.
Travis is failing with cppcheck on `activate_app_request_test.cc` file that I didn't touch.

shoamano83 commented:
Rebased once again.

iCollin commented:
Hello @shoamano83 , we are preparing to implement this change. Do you mind rebasing it again?

shoamano83 commented:
Hi @iCollin , OK let me try to rebase once again.

shoamano83 commented:
I've rebased onto latest `develop` branch. Please kindly review.
Thanks.

iCollin commented:
@shoamano83 Can you please explain why you chose to implement in such a way that audio_pass_thru_capabilities_ can be either an array or map? I imagine it might save some headache in the future if this is implemented as an array in all cases. When only one option is defined in the capabilities it could be an array of length one.

shoamano83 commented:
Hi @iCollin , that's just because the user of `audio_pass_thru_capabilities_`, which is `FillUIRelatedFields()` in register_app_interface_request.cc, already has a logic of switching between a map and array. So to align with the logic, I thought I should return a map in the case of old format.
It shouldn't be a problem to always make `audio_pass_thru_capabilities_` an array, let me try to work on it.

shoamano83 commented:
Hi @iCollin , I updated my implementation. Could you kindly review again? Thanks!

iCollin commented:
@shoamano83 Thank you for the update. I found a few things that do need to be addressed:

* The unit test `SetAudioPassThru_SUCCESS` is currently failing
* `audio_pass_thru_capabilities_` can still be set to a MAP in `ui_get_capabilities_response.cc`
* `register_app_interface_request.cc` should be able to assume `audio_pass_thru_capabilities_` is an array

shoamano83 commented:
Thanks @iCollin .

> The unit test SetAudioPassThru_SUCCESS is currently failing

I didn't detect this issue when I updated my PR last week. I pushed another commit now, and I did a clean build and ran `make test` to make sure all tests passed. If you still see the issue, could you kindly share error logs?

> audio_pass_thru_capabilities_ can still be set to a MAP in ui_get_capabilities_response.cc

Okey, updated with commit d46fd33311088f4d3b0b42a381feb28a4f23ad50.

> register_app_interface_request.cc should be able to assume audio_pass_thru_capabilities_ is an array

Yeah but this seems like a logic change that I didn't propose in my proposal. I prefer to keep `FillUIRelatedFields` supporting both array and non-array for backward compatibility and also for avoiding confusion among developers.
Let me know if I didn't get your message correctly.

iCollin commented:
Thank you @shoamano83 , I see all tests passing now. 

> I prefer to keep FillUIRelatedFields supporting both array and non-array for backward compatibility and also for avoiding confusion among developers.

With the changes in https://github.com/smartdevicelink/sdl_core/commit/d46fd33311088f4d3b0b42a381feb28a4f23ad50 , this if statement will always evaluate to true.

`if (smart_objects::SmartType_Array == hmi_capabilities.audio_pass_thru_capabilities()->getType())`

Therefore I suggest we remove the if statement and else block completely. I believe it would cause more confusion for developers, but if you disagree please do explain.

shoamano83 commented:
Hi @iCollin , my concern is that the change isn't included in the proposal. Developers may be confused of a code change that hasn't been described in the proposal. Maybe we could create another PR (like enhancement PR or refactor PR) for that change?

I'm pushing a commit that includes the change, so you can approve and include it if you want.
Thanks.

iCollin commented on src/components/application_manager/rpc_plugins/sdl_rpc_plugin/src/commands/hmi/ui_get_capabilities_response.cc:
```diff
@@ -73,9 +73,24 @@ void UIGetCapabilitiesResponse::Run() {
     hmi_capabilities.set_audio_pass_thru_capabilities(
-        msg_params[strings::audio_pass_thru_capabilities]);
+        msg_params[strings::audio_pass_thru_capabilities_list]);
+  } else if (msg_params.keyExists(strings::audio_pass_thru_capabilities)) {
```
Sorry for the late comment, but in this RAI response `msg_params[strings::audio_pass_thru_capabilities]` will always be an object as defined in the `HMI_API`. If you wouldn't mind just removing the case for processing it as an array we will be ready to merge. Thank you!
